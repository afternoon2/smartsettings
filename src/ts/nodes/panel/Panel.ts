import cuid from 'cuid';

import { ParentNode, AnyControl } from '../parent/ParentNode';
import { PanelProps, PanelPosition, SectionOptions } from '../nodes.types';
import { InternalState } from '../../root/RootNode';

import Base from '../../../sass/base.sass';
import Styles from '../../../sass/panel.sass';
import { ControlListener, ControlListenerUpdate, ControlOptions } from '../../controls/control/Control.types';
import { SectionNode } from '../section/Section';

export class PanelNode extends ParentNode {
  public bodyElement: HTMLElement;
  public headerElement: HTMLElement;
  private toggleElement: HTMLAnchorElement;
  private nameElement: HTMLParagraphElement;

  protected static template = (state: InternalState): string => `<header
      class="${Styles.panel__header}"
      ${state.draggable ? 'draggable="true"' : ''}
    >
      <p class="${Styles.panel__name}" title="${state.name}">
        ${state.name}
      </p>
      <a class="${Styles.panel__link}">
        ${state.collapsed ? '&#9662;' : '&#9656;'}
      </a>
    </header>
    <main class="${Styles.panel__body} ${state.collapsed ? Base.hidden : ''}">
    </main>`;

  constructor(props: PanelProps) {
    super(props, PanelNode.template);
    this.fillInElement(Styles.panel);
    this.headerElement = this.element.querySelector(`.${Styles.panel__header}`) as HTMLElement;
    this.nameElement = this.headerElement.querySelector('p') as HTMLParagraphElement;
    this.toggleElement = this.headerElement.querySelector('a') as HTMLAnchorElement;
    this.bodyElement = this.element.querySelector(`.${Styles.panel__body}`) as HTMLElement;
    this.listeners.set('top', this.onPositionChanged);
    this.listeners.set('left', this.onPositionChanged);
    if (props.listener) {
      this.listeners.set('panel', props.listener);
    }
    this.bindEventListeners();
    if (!this.state.top || !this.state.left) {
      this.setPosition({
        top: 10,
        left: 10,
      });
    }
  }

  get position(): PanelPosition | null {
    return this.state.top && this.state.left ?
      {
        top: this.state.top as number,
        left: this.state.left as number,
      } :
      null;
  }

  setPosition(position: PanelPosition) {
    this.state.top = position.top;
    this.state.left = position.left;
  }

  setListener(listener: ControlListener) {
    this.listeners.set('panel', listener);
  }

  destroy() {
    this.registry.clear();
    this.listeners.clear();
    this.parentElement.removeChild(this.element);
  }

  remove(name: string) {
    const toRemove = Array.from(this.registry.values())
      .find((el: AnyControl | SectionNode) => el.name === name);
    if (!toRemove) {
      this.registry.forEach((value: AnyControl | SectionNode) => {
        if (value instanceof SectionNode) {
          value.remove(name);
        };
      });
    } else {
      this.bodyElement.removeChild(toRemove.element);
      this.registry.delete(toRemove.id);
    }
  }

  removeById(id: string) {
    const localElement = this.registry.get(id);
    if (localElement) {
      this.bodyElement.removeChild(localElement.element);
      this.registry.delete(id);
    } else {
      this.registry.forEach((value: AnyControl | SectionNode) => {
        if (value instanceof SectionNode) {
          value.removeById(id);
        }
      })
    }
  }

  removeAll() {
    while (this.bodyElement.firstChild) {
      this.bodyElement.removeChild(
        this.bodyElement.firstChild,
      );
    }
    this.registry.clear();
  }

  control(
    control: string,
    name: string,
    options: ControlOptions | null,
    listener?: ControlListener,
  ): AnyControl | null {
    return this.createControl(
      control,
      name,
      options,
      {
        control: listener,
        panel: this.listeners.get('panel'),
      },
    );
  }

  section(
    name: string, 
    options: SectionOptions, 
    listener?: ControlListener
  ): SectionNode {
    const id: string = cuid();
    const section = new SectionNode({
      id,
      name,
      options,
      parentElement: this.bodyElement,
      listener,
      panelListener: this.listeners.get('panel'),
    });
    this.registry.set(id, section);
    return section;
  }

  private bindEventListeners() {
    this.toggleElement.addEventListener('click', this.toggleClickListener);
    this.nameElement.addEventListener('click', this.toggleClickListener);
  }

  private toggleClickListener = (event: Event) => {
    event.preventDefault();
    if (this.state.collapsed === true) {
      this.open();
    } else {
      this.close();
    }
  }

  private onPositionChanged: ControlListener = (update: ControlListenerUpdate) => {
    const cssProp: string = update.key;
    // @ts-ignore
    this.element.style[cssProp] = `${update.value}px`;
  }
}
