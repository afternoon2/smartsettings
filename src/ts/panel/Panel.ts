import { RootNode } from '../root/RootNode';
import { ControlListener } from '../controls/controls.types';

import Base from '../../sass/base.sass';
import Styles from '../../sass/panel.sass';

export type PanelOptions = {
  collapsed?: boolean,
  draggable?: boolean,
  locked?: boolean,
  invisible?: boolean,
};

export class Panel extends RootNode {
  public parentElement: HTMLElement = document.body;
  public element: HTMLDivElement;

  constructor(
    title?: string,
    parent?: HTMLElement | string,
    options?: PanelOptions,
    listener?: ControlListener | null,
    id?: string | null,
  ) {
    super(
      title || 'SmartSettings @2.0.0',
      options || {},
      listener,
      id,
    );
    this.assignParentElement(parent);
    this.element = this.createPanelNode();
    this.parentElement.appendChild(this.element);
  }

  private assignParentElement(parent: HTMLElement | string | undefined) {
    if (parent && this.isValidElement(parent)) {
      this.parentElement = this.getParentElement(parent);
    }
  }

  private createPanelNode(): HTMLDivElement {
    const { locked, invisible, draggable, collapsed } = this.options as PanelOptions;
    const { id, name } = this;
    const rootNode: HTMLDivElement = document.createElement('div');
    rootNode.classList.add(Styles.panel);
    if (locked) {
      rootNode.classList.add(Base.locked);
    }
    if (invisible) {
      rootNode.classList.add(Base.hidden);
    }
    if (draggable) {
      rootNode.setAttribute('data-draggable', "true");
    }
    rootNode.setAttribute('id', id);
    const template: string = `
      <header class="${Styles.panel__header}">
        <p class="${Styles.panel__name}" title="${name}">
          ${name}
        </p>
        <a class="${Styles.panel__link}">
          ${collapsed ? ' &#9662;' : '&#9656'}
        </a>
      </header>
      <main class="${Styles.panel__body} ${collapsed ? Base.hidden : ''}">
      </main>
    `;
    rootNode.insertAdjacentHTML('beforeend', template);
    return rootNode;
  }
}
