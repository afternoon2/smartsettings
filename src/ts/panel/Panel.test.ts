import { Panel, PanelOptions } from './Panel';

describe('Panel', () => {
  test('If it initializes correctly', () => {
    const panel = new Panel(
      'Panel',
      null,
      {
        locked: true,
        collapsed: true,
        draggable: true,
      },
      null,
      null,
    );
    panel.id = 'id';
    expect(panel).toBeInstanceOf(Panel);
    expect(panel.parentElement).toBe(document.body);
    expect(panel.element).toBeInstanceOf(HTMLDivElement);
  });

  test('If it mounts to the DOM correctly', () => {
    const panel = new Panel(
      'Panel',
      null,
      {
        locked: true,
        collapsed: true,
        draggable: true,
      },
      null,
      'test_id',
    );
    const element = <HTMLDivElement>panel.element;
    expect(element).toMatchSnapshot();
  });

  test('If it invokes global listener on changes correctly', () => {
    const listener = jest.fn();
    const panel = new Panel('Panel', null, null, listener, null);
    panel.id = 'id';
    (<PanelOptions>panel.options).invisible = true;
    (<PanelOptions>panel.options).collapsed = true;
    (<PanelOptions>panel.options).locked = true;
    expect(listener).toHaveBeenCalledTimes(3);
  });
});
