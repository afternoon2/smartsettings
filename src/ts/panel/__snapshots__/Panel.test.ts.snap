// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Panel If it mounts to the DOM correctly 1`] = `
<div
  class="panel locked"
  data-draggable="true"
  id="test_id"
>
  
      
  <header
    class="panel__header"
  >
    
        
    <p
      class="panel__name"
      title="Panel"
    >
      
          Panel
        
    </p>
    
        
    <a
      class="panel__link"
    >
      
           â–¾
        
    </a>
    
      
  </header>
  
      
  <main
    class="panel__body hidden"
  >
    
      
  </main>
  
    
</div>
`;
